### ISSUE TITLE
User profile attributes should only accept a single value unless configured otherwise

### ISSUE THREAD

**Issue by jonkoops** [keycloak/keycloak#23539]
When a user profile attribute is configured it is still possible to give it a multi-valued string. For example, if an attribute named `age` is defined the following are all valid representations of that attribute on the user:

```json
{
  "attributes": { "age": ["18"] }
}
```
```json
{
  "attributes": { "age": ["18", "42"] }
}
```

This makes the value of the attribute extremely ambiguous for whomever is consuming this data from the REST API, making it had to reason about. One cannot make assumptions about which of these values is correct, allowing room for bugs to be introduced.

Therefore the proposal is as follows:
- Add a boolean property called `isMultiValued` to [`UserProfileAttribute`](https://github.com/keycloak/keycloak/blob/22123e0afc39935a88306deb816d0146519bf21e/services/src/main/java/org/keycloak/userprofile/config/UPAttribute.java#L28), defaulting to `false`.
- If  `isMultiValued` is `true` an array of multiple strings is accepted as a value (e.g. `"age": ["18", "42"]`).
- If `isMultiValued` is `false` an array of a single string accepted as a value (e.g. `"age": ["18"]`).
- A checkbox will be added to the form in the admin console to toggle the value of `isMultiValued`.

**Comment by pedroigor** [comment]
@jonkoops Can we close this one in favor of https://github.com/keycloak/keycloak/issues/19606?

That one covers both UI and API ends.

**Comment by jonkoops** [comment]
> Can we close this one in favor of #19606?

@pedroigor Since this one is a bit more implementation specific I think it would be better if we close the other issue. This one also covers the UI regardless.

**Comment by jonkoops** [comment]
Ok, so let me try to break down your comments here.

> To communicate UI-related information [1] with clients we have some built-in `annotations`. The annotations are basically supported by Freemarker templates. We don't necessarily need to have a specific `multivalued` property if we can actually communicate to clients how the attribute should actually be rendered.

So this is kinda hard, as you really can't know in advance where these attributes will be consumed. It might be correctly implemented in our case, but that doesn't mean that it is logical for 3rd party consumers of this data. It is also perfectly valid at the moment to create a user profile attribute without any annotations or validations. Making it practically impossible to determine this accurately.

I am going to operate on the assumption that you referring here to the 'input type' being the specifier for applications to understand whether this is a multivalue or not. This has several distinct problems:

### It's ambiguous about whether a field is multi-valued
For example, imagine we had a user profile attribute called `phoneNumber` with an input type set to `html5-tel`:

![image](https://github.com/keycloak/keycloak/assets/695720/6d75b39e-e388-417b-9d86-474516c7ccd1)

This would create the following user profile configuration for the client to consume:

```json
{
  "name": "phoneNumber",
  "displayName": "Phone Number",
  "required": false,
  "readOnly": false,
  "annotations": {
    "inputType": "html5-tel"
  }
}
```

With something like the following attribute data backing it up:

```json
{
  "attributes": {
    "phoneNumber": [
      "06123456"
    ]
  }
}
```

There is no way for a consumer of this API to know whether this needs to be rendered as a single field. Or if it would be possible to add multiple phone numbers. In fact, we will just have to assume that it's either always a single field, or multi-valued. In the case of the Account Console we simply render out a single input for this field:

![image](https://github.com/keycloak/keycloak/assets/695720/ce6cffee-4ca8-42e4-9b1a-3d6822fd2ee3)

So if we wanted to accomplish having a single attribute with multiple values in the form (e.g. a new input for each value), there is currently no way of doing so.

So either we need to have a new input type of `html5-tel-multivalue`, or we need a new property on the attribute itself. For example, if we had the `multiValued` property:

```json
{
  "name": "phoneNumbers",
  "displayName": "Phone Numbers",
  "required": false,
  "readOnly": false,
  "multiValue": true,
  "annotations": {
    "inputType": "html5-tel"
  }
}
```

Now it is explicitly clear to the API consumer that `phoneNumbers` can contain multiple values, which would allow us to build forms like this:

![image](https://github.com/keycloak/keycloak/assets/695720/5c4ce9ff-d2d3-468b-874a-27ff157931d9)

### There is no validation if a single or multiple values are set

Even if I set an input type (such as `html5-tel`) in the earlier example. There is no validation that guarantees that only a single value is set. So in our earlier example, even if I would assume a an attribute is only backed by a single field in the form, a multi-value can still be set anywhere:

```json
{
  "attributes": {
    "phoneNumbers": [
      "+13469763709",
      "+15736337698"
    ]
  }
}
```

In this case, which is the correct value? The first, the last? It's just not possible to really tell. People make mistakes, and things can _and will_ be set in this way by mistake. By making it explicit if something is multi-value, we can apply validations when users attempt to set invalid values.

### Annotations are not required

Even if we discard all the aforementioned problems, we're still left with another: Annotations are not required to create a user attribute.  This means that in the worst case scenario the client only has the following metadata to go on:

```json
{
  "name": "phoneNumbers",
  "displayName": "Phone Numbers",
  "required": false,
  "readOnly": false
}
```

With this information, essentially nothing valid can be determined and we will likely just have to assume this is a multi-valued text.

I hope this elaborates as to why I think we need to be able to configure attributes as multi-value (or not) and ensure that data is validated correctly.

**Comment by jonkoops** [comment]
@antikalk I think the discussion above might also be relevant for you, considering you logged the original issue asking for support for this feature.

**Comment by pedroigor** [comment]
@jonkoops I'm not talking about the built-in `inputType` annotation necessarily. As I mentioned, you can always have your own.

If we choose another annotation for the admin console that would probably be documented so that 3rd-party admin clients can operate in the same way.

**Comment by antikalk** [comment]
@jonkoops thanks for the examples, this is exactly what I am looking for! 👍 

@pedroigor imo it would also make total sense to have it in an annotation, to indicate the intend that this field is multivalued. 

But, please correct me if I am wrong, annotations do not have any impact on how the server side validates the input, right? E.g. if I am using html5-tel inputType it is not validated on the server side that the input is really a phone number. You would have to add a validator for that additionally. Which I don't think is too much of an issue - as this can just simply be configured once.

But for the multivalue this would become tricky: As the default would be `"multivalued": false` or the annotation not being set at all you would have to add a "singleValueValidator" to every field that is not multivalued if you want to make sure that not multiple values are stored.

**Comment by pedroigor** [comment]
> But, please correct me if I am wrong, annotations do not have any impact on how the server side validates the input, right?

Yeah, the server is blind about the annotations set to an attribute. It adds no value to the processing of the user profile and its attributes.

> As the default would be "multivalued": false or the annotation not being set at all you would have to add a "singleValueValidator" to every field that is not multivalued if you want to make sure that no multiple values are stored.

It is also tricky when you have `multivalued: true` and now you change the attribute to be single-valued. I'm not sure what the server should do in this case. If we want to add validation for `multivalued`, IMO that would be a built-in validation. Therefore, the proposal from @jonkoops to make it a first-class citizen property in the attribute metadata makes sense.

However, as I mentioned before, we are also going to face other tricky situations. For instance, if you create a field as `multivalued` and then switch to single-valued, what the server should do? Truncate the values? Which values the server should ignore/remove? Does the admin need to first remove values and then switch to single-valued?

That is why I think we can start simple and rely on annotations as much as possible to drive UI rendering. Completely decoupling the server (to which attributes are always multivalued) from the client. In your example, we could have something in the UI that automatically adds/removes a validator depending on how the `multivalued` attribute is set.

**Comment by jonkoops** [comment]
@pedroigor I agree with your assessment that this is definitely an enhancement we can do later. Let's wait until after we have stabilized User Profile as a feature before we start to consider what the implementation for this should be.

**Comment by mposolda** [comment]
+1 for something like having `multivalued` property on user-profile configuration for individual attributes.

Especially since some of our validators are validating just the 1st value - like for example `RegistrationUsernameExistsValidator` calling `String value = values.get(0);` and working just with the very 1st value and ignoring the other values. And ignoring all the values besides the 1st can have some unexpected side-effects...

I can see that migration can be the concern as mentioned already. For example when someone switches attribute `foo` from `multivalued: true` to `multivalued: false`, but attribute `foo` already has multiple values for user `john`. I think that when this is switched, we should enforce that update of user `john` won't be allowed (either by Account REST API or by admin REST API) if the JSON sent in the UPDATE request contains multiple values of attribute `foo`. This will give user/admin to adapt to the change and make sure that he is able to update user just with valid content (so single value for attribute `foo`).
I suppose that for our UI layers (both Account UI and admin UI), we can have annotations, which will allow only to make sure that single value is rendered for such attribute and hence also update request will contain just the single value, which was rendered in UI? So update of such user will be easily possible unless I am missing something.

**Comment by jonkoops** [comment]
> I suppose that for our UI layers (both Account UI and admin UI), we can have annotations, which will allow only to make sure that single value is rendered for such attribute and hence also update request will contain just the single value, which was rendered in UI? So update of such user will be easily possible unless I am missing something.

Yeah, I think we would simply truncate the existing value if an attribute is not `multivalued`, but has multiple values assigned to it. We already have logic in the front-end to [determine an input type](https://github.com/keycloak/keycloak/blob/135a5c9b0d930d5086467b8093ad1e5eddd59f9b/js/libs/ui-shared/src/user-profile/UserProfileFields.tsx#L215-L240) based on the length of the values, but this of course quite a bit more error prone to determine than an explicit field.

**Comment by pedroigor** [comment]
@jonkoops After talking with @mposolda, we can continue the discussion about this one once we deliver the feature. We also need to review server-side code because we lack consistency when handling attribute values. For instance, https://github.com/keycloak/keycloak/issues/23539#issuecomment-1835653589.

We are thinking about moving this one to a follow-up epic.

**Comment by ahus1** [comment]
I just tried out user profiles for a demo, and I found the following behavior in the admin UI when changing a newly added user attribute: 

* If the database already stores a multi-value field with two entries, in the admin UI I can remove that second entry and save it. 
* Once I removed everything but the last entry, the UI changes and I can no longer add additional values. 

I didn't add any validations or attributes, as I didn't find any one that looked to me like it would influence this behavior.

I then searched the open issues and found that seems to relate to multivalued fields, and that it is no longer planned for KC24, which surprises me as users won't be able to undo changes as described above. 

So I wonder if the the behavior I describe in this comment is related to this issue, or if it should be a new bug.

**Comment by pedroigor** [comment]
@ahus1 Can you please open a new bug? I think it is somewhat related but it does not necessarily mean we need to introduce changes to the configuration schema or the SPI.

**Comment by jonkoops** [comment]
This bug was already reported several times, I don't think it can be fixed unless a concept of explicit multi-value is introduced. The algorithm used to determine whether a field is multi-valued or not on the client side is non deterministic, and needs to infer this from a lack of detail. We can keep changing the default behaviour all we want (we already had 4 PRS to do so), but it will always be a workaround and result in undefined and unwanted behaviour.

**Comment by ahus1** [comment]
@pedroigor - given @jonkoops's comment, I won't open a new issue. Still, I think this should be considered high priority (a blocker?) for user attribute handling in KC24 as handling multi-values attributes is IMHO not working before this is implemented

I think is a regression compared to previous versions of Keycloak unless someone can describe me how users will be able to work around this.

cc: @mposolda

**Comment by pedroigor** [comment]
After talking with @ahus1, looks like the best is indeed add the `multivalued` field to the configuration. One of the main arguments is that the multivalued setting is orthogonal to the input type. Therefore, using annotations to render a field as multivalued isn't enough because you have different input types that can be multivalued.

So, the idea is to:

* Add a `multivalued` setting to the configuration
* Change the Attribute UI to allow setting this prperty
* Review UIs (admin, templates, account) to support multivalued fields

I'll provide something next week.

**Comment by jonkoops** [comment]
Nice, let me know if you need some assistance with this on the UI side @pedroigor
